src/server/services/wms/qgswmsrestorer.h

```diff
@@ -64,4 +65,32 @@ class QgsLayerRestorer
     QMap<QgsMapLayer *, QgsLayerSettings> mLayerSettings;
 };
 
+namespace QgsWms
+{
+  /**
+   * \ingroup server
+   * RAII class to restore the rendering context configuration on destruction
+   * \since QGIS 3.14
+   */
+  class QgsWmsRestorer
+  {
+    public:
+
+      /**
+       * Constructor for QgsWmsRestorer.
+       * \param context The rendering context to restore in its initial state
+       */
+      QgsWmsRestorer( const QgsWmsRenderContext &context );
+
+      /**
+       * Default destructor.
+       */
+      ~QgsWmsRestorer() = default;
+
+    private:
+
+      std::unique_ptr<QgsLayerRestorer> mLayerRestorer;
```

Comment:
Does this need to be a unique ptr? It could just be a value instead.

Reply:
The `QgsLayerRestorer` constructor needs a list of layers.

Reply:
Yes but if you could initialise it:

QgsWmsRestorer::QgsWmsRestorer( const QgsWmsRenderContext &context )
    : mLayerRestorer( context.layers() )
{}

Reply:
Oh yes of course, silly me :upside_down_face:.

Thanks Nyall, I'm going to do that.